### PR TITLE
Adding post from carlos with title 'Imposing invariants in Scala'

### PR DESCRIPTION
**Pull request by cmhteixeira**

### CODE REVIEW

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+sealed trait Seat
+
+case class Seated(row: Int, column: Int) extends Seat
+case class Tribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />
```
Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: can you put this image into `public` folder of the repo?

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+The first thing to notice is that the above does not deliver what we want. The following compiles:  
+
+```scala
+val reservation = Reservation(
+  usertTicket = Ticket("id1", 123, VipTribuneSeat("east")),
+  guestTicket = Ticket("id2", 123, Standing)
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribuneSeat</code>. The compiler is satisfied since this a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> only line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!
```
Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: The compiler is satisfied since this **is** a sub-type
missing `is`

Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Bad news since type <code>SeatType</code> **only** line

This `only` is confusing, did you mean `on` ?

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    case _ => false
+  }
+
+  require(verifyTypeRestriction)
+}
+```
+
+On this approach we delegate to the users of case class Reservation the responsibility of never trying to instantiate a value that does not conform to the rules. If they do breach the rule, the code compiles but throws an exception at run time.  
+
+While this is the least adequate of the solutions, we can the peace of mind that, if it does blow up, it will do so on instantiation of the object. Meaning that some other block of code further down can expect the invariant to be respected and avoid dealing with the inconsistent cases (since if the block of code is being executed, the previous instantiation succeed). We have centralized the point of failure to the instantiation of an object.
```
Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: we can **____** peace of mind
I think `the` does not fit here, try some verb `have` `get` `find`...

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+
+```scala
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+)
+```
+the compiler can impose <code>SeatInfo</code> to have as upper bound the type <code>Seat</code>. It cannot however impose <code>SeatInfo</code> to be one of the sub-types exclusively. Meaning that upon userTicket: <code>Ticket[VipTribune]</code> and guestTicket: <code>Ticket[NormalSeat]</code>, both <code>VipTribune</code> and <code>NormalSeat</code> can pretend to be type <code>Seat</code> which would result in userTicket: <code>Ticket[Seat]</code> and guestTicket: <code>Ticket[Seat]</code> which in turn conforms with the upper bound constraint (Reservation[SeatInfo <: Seat]). The issue really is that a given type A is a subtype of itself.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />
```
Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: can you put this image in public folder of the repo?

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+The trick on this approach is to use a generalized type constraint. These are not built-in features of the language (like type bounds) but rather smart ways some smart people have engineered to leverage the type system of Scala to accomplish some more impressive things.
+
+
+There are 2 or 3 generalized type constraints on the standard library of Scala on package scala.Predef which is imported by default. One of the most famous is <code> <:< </code>.
+
+The two we are going to use are found on the external library Shapeless by Miles Sabin.
+
+The first is <code>=:!=[A, B]</code>. This can also be written <code>A =:!= B</code>. 
+
+I will not try to explain the intricacies of these things. A detailed explanation can be found at #here.
```
Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: `at #here`

missing link?

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+[error]  both method neqAmbig1 in package shapeless of type [A]=> A =:!= A
+[error]  and method neqAmbig2 in package shapeless of type [A]=> A =:!= A
+[error]  match expected type com.github.cmhteixeira.invariants.Seat =:!= com.github.cmhteixeira.invariants.Seat
+[error] val foo = Reservation(
+[error]                           ^
+```
+
+The error is not very helpful, but one must live with it. (footnote: answer by Régis Jean-Gilles on https://stackoverflow.com/questions/6909053/enforce-type-difference)  
+
+This approach, similarly to the 2 other approaches, suffers from an inconvenient. The type SeatInfo on a given reservation is lost at run time due to type erasure. We cannot pattern match on a given reservation as on the following example:
```
Comment (nbardiuk) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: suffers from an inconvenient **____**

missing a noun, `inconvenient` should describe something

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribuneSeat</code>. The compiler is satisfied since this a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> only line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  
+Before launching an error, the compiler realizes both <code>Standing </code> and <code>VipTribuneSeat</code> are actually of sub-types of <code>Seat</code> ; this means they can pretend to be their super-type (<code>Seat</code>). Because <code>Seat</code> respects the type bound <code>Seat</code> <:  <code>Seat</code>, the compilation is successful with val reservation: Reservation[Seat].
+
+<strong> There are at least 4 ways to enforce the restriction of the opera-house: </strong>
+
+
+1. using require(condition) on the class constructor of Reservation  
+2. Hiding the public constructor of the class, defining a public one that validates and returns Option[Reservation]
```
Comment (Amuxix) [inline _posts/2018-12-04-imposing-invariants-in-scala.md:66]: I don't think this works, there is a bug related to changing the visibility of the main constructor in case classes, even you set it to private the generated `apply` method will still be visible.

Below you don't use a case class but upon reading this seems that you will attempt to hide the constructor on a case class.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 1 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: I have uploaded the image, but left the code still pointing to wordpress. I only know how to link to the enear's blog repo after the blog post is deployed. (i will change it then).  
Or maybe there is some relative path I can use ?

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 2 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done

Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: Yes.
Done

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 3 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 4 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: See first comment

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 5 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+){
+
+  private def verifyTypeRestriction: Boolean = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => true
+    case (VipTribuneSeat(_), VipTribuneSeat(_)) => true
+    case (_: Standing.type , _: Standing.type ) => true
+    case _ => false
```
Comment (Amuxix) [inline _posts/2018-12-04-imposing-invariants-in-scala.md:83]: You do the checks above in a somewhat weird way, you unapply the classes(you could match the types like you do for `Standing`) and on the object you can do ``(`Standing`, `Standing`)``

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 6 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: inconvenient can actually be used as a noun as well.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribuneSeat</code>. The compiler is satisfied since this a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> only line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  
+Before launching an error, the compiler realizes both <code>Standing </code> and <code>VipTribuneSeat</code> are actually of sub-types of <code>Seat</code> ; this means they can pretend to be their super-type (<code>Seat</code>). Because <code>Seat</code> respects the type bound <code>Seat</code> <:  <code>Seat</code>, the compilation is successful with val reservation: Reservation[Seat].
+
+<strong> There are at least 4 ways to enforce the restriction of the opera-house: </strong>
+
+
+1. using require(condition) on the class constructor of Reservation
```
Comment (Amuxix) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: `using` should be capitalized.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+Let us recall the original problem. On:  
+
+```scala
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+)
+```
+the compiler can impose <code>SeatInfo</code> to have as upper bound the type <code>Seat</code>. It cannot however impose <code>SeatInfo</code> to be one of the sub-types exclusively. Meaning that upon userTicket: <code>Ticket[VipTribune]</code> and guestTicket: <code>Ticket[NormalSeat]</code>, both <code>VipTribune</code> and <code>NormalSeat</code> can pretend to be type <code>Seat</code> which would result in userTicket: <code>Ticket[Seat]</code> and guestTicket: <code>Ticket[Seat]</code> which in turn conforms with the upper bound constraint (Reservation[SeatInfo <: Seat]). The issue really is that a given type A is a subtype of itself.
```
Comment (Amuxix) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Here you say: <code>NormalSeat</code> can pretend to be type <code>Seat</code>, I think a more standard way of expressing this(using OOP terms) is that `NormalSeat`can be viewed as a `Seat`. This is a pretty minor change I guess.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 1 above)
Comment (nbardiuk) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: you can use `type hierarchy`

browser adds domain to the url automatically

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 11]
```diff
@@ -0,0 +1,359 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please replace the "T" in `The buyer` with a lower case "t", as that is not the beginning of a new sentence.
Also here `of the reservation himself, and his guest.` you can remove the comma.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 12]
```diff
@@ -0,0 +1,359 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.  
+As in any show-house, there are many types of seats; some are standing, some normal and some on vip tribunes.
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please put `vip` in upper case since "VIP" acronym is usually written in upper case.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+
+
+
+As a restriction from the opera house, both buyer and guest must be on the same seating area. For example, it is not permitted for one ticket to be standing while the other for the normal seated area.  
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please remove the extra whitespace before the question mark.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+
+```scala
+sealed trait Seat
+
+case class NormalSeat(row: Int, column: Int) extends Seat
+case class VipTribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: I understand what you mean here, but the sentence has a weird construction in English. It makes sense in Portuguese.
Right now, I don't have any suggestion. I'll come back to this later.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+The first thing to notice is that the above does not deliver what we want. The following compiles:  
+
+```scala
+val reservation = Reservation(
+  usertTicket = Ticket("id1", 123, VipTribune("east")),
+  guestTicket = Ticket("id2", 123, Standing)
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribune</code>. The compiler is satisfied since this is a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> on line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: I'd replace 
> From the user ticket on line 2 it thinks it is <code>VipTribune</code>.

with `Regarding user ticket on line 2, it thinks it is <code>VipTribune</code>.` 

Also here
> Bad news since type <code>SeatType</code> on line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  

I'd would rephrase the last part to `on line 3 refer to, well, the same type! ` -> moved the "to" so it's before the "well".

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+
+
+```scala
+sealed trait Reservation[SeatType <: Seat] {
+  val userTicket: Ticket[SeatType]
+  val guestTicket: Ticket[SeatType]
+}
+```
+
+and on the companion object we have the actual implementation:
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please capitalize the "a" in "and" at the beginning of the sentence, since you're starting a new one.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+sealed trait Seat
+
+case class NormalSeat(row: Int, column: Int) extends Seat
+case class VipTribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />
```
Comment (ragb) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Don't like the colors.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+<strong> There are at least 4 ways to enforce the restriction of the opera-house: </strong>
+
+
+1. Using require(condition) on the class constructor of Reservation  
+2. Hiding the public constructor of the class, defining a public one that validates and returns Option[Reservation]
+3. Using generalized type constrains
+4. Using OOP - subtyping
+
+
+# 1. Using require(condition) on the class constructor of Reservation
```
Comment (ragb) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please use level 3 headings or something, level 1 should be only used for the post / page title.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+  def apply[SeatType <: Seat](userTicket: Ticket[SeatType], guestTicket: Ticket[SeatType]): Option[Reservation[SeatType]] = (userTicket.seat, guestTicket.seat) match {
+    case (NormalSeat(_, _), NormalSeat(_, _)) => Some(ReservationImpl(userTicket, guestTicket))
+    case (VipTribune(_), VipTribune(_)) => Some(ReservationImpl(userTicket, guestTicket))
+    case (_: Standing.type , _: Standing.type ) => Some(ReservationImpl(userTicket, guestTicket))
+    case _ => None
+  }
+}
+```
+
+Notice the relevance of declaring the trait sealed. It prevents the user from extending the trait by another class on some other file, which would allow him to bypass the factory method <strong>apply</strong> above. Equally important is to have ReservationImpl as private so that it cannot be extended on client code. In fact, I guess we could even ignore ReservationImpl altogether and instantiate an anonymous class instead.
```
Comment (ragb) [inline _posts/2018-12-04-imposing-invariants-in-scala.md:159]: good point here. One oter thing you may consider (not specially here but in some places else) is to make your case classes final to avoid any kind of external subtyping.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 20]
```diff
@@ -0,0 +1,359 @@
...
+
+It is written <code>=:!=[A, B]</code>; this is a parameterized trait named <code>=:!=</code> with two types. This can also be written <code>A =:!= B</code>. 
+
+I will not try to explain the intricacies of these things. A detailed explanation on related type constraints can be found <a id="union-types-ref" href="#union-types">here</a>. I hope to transmit a feeling on how to use them. 
+
+Each of these type constraints is used with a specific goal. The goal of <code>A =:!= B</code> is to impose that types <code>A</code> and <code>B</code> are different, whatever they might be. This is accomplished via implicits. Specifically, you declare an implicit variable of type <code>A =:!= B</code>, where <code>A</code> and <code>B</code> are some two types you want to make sure are different. I would say that normally, either <code>A</code>, <code>B</code> or both are abstract/parameterized types, like type <code>Ole</code> below.
+
+```scala
+import shapeless.=:!=
+def foo[Ole](a: Ole)(implicit ev: =:!=[Ole, Int]): A = a
```
Comment (ragb) [inline _posts/2018-12-04-imposing-invariants-in-scala.md:196]: I'd change this to the infix notation: it's more readable and what you generally found on the wild.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 6 above)
Comment (nbardiuk) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: agree, as a noun it needs different ending `inconvenience`

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+non-variable type argument com.github.cmhteixeira.invariants.Standing.type in type pattern com.github.cmhteixeira.invariants.Reservation[com.github.cmhteixeira.invariants.Standing.type] is unchecked since it is eliminated by erasure
+[warn]     case a: Reservation[Standing.type] => "Go to gate 1"
+```
+
+Fortunately the type parameter SeatInfo on Reservation is not a phantom type. It is associated with the value of seat within each Ticket (userTicket and guestTicket) of the reservation. To avoid the above problem we must therefore pattern match on the seats of each ticket.  
+
+```scala
+def guideUsersToSeat(reservation: Reservation[Seat]): String = {
+    val userSeat: Seat = reservation.userTicket.seat
+    val guestTicket: Seat = reservation.guestTicket.seat
```
Comment (Amuxix) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: I think this should named `guestSeat` not `guestTicket`

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 14 above)
Comment (Amuxix) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: I agree with Sofia, I believe a better way to express this would be `This DSL might be easier to understand with the following diagram:`

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 17 above)
Comment (Amuxix) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: I think the arrows are the wrong way, they should be pointing away from `Seat` as they extend `Seat` and not the other way round.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 6 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: It seems you are right.  I will change.
Funny, I was very sure that `inconvenient` could also be used as a noun.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 11 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 12 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 13 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 16 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 17 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: See answer from **artamonovdev** at https://stackoverflow.com/questions/1874049/explanation-of-the-uml-arrows.
Also guys on the office confirm. :)

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 18 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: ok. Changed to level 2 headings.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 19 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md:159]: didn't remember this.  Thanks.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 20 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md:196]: infix notation would be another source of confusion for people to whom evidences and this kind of stuff is already new.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+[error]  both method neqAmbig1 in package shapeless of type [A]=> A =:!= A
+[error]  and method neqAmbig2 in package shapeless of type [A]=> A =:!= A
+[error]  match expected type com.github.cmhteixeira.invariants.Seat =:!= com.github.cmhteixeira.invariants.Seat
+[error] val foo = Reservation(
+[error]                           ^
+```
+
+The error message is not very helpful, but one must live with it <a id="SO-answer-error-message-union-types-ref" href="#SO-answer-error-message-union-types"> ¹ </a>
+
+This approach, similarly to the 2 other approaches, suffers from an inconvenient. The type <code>SeatType</code> on a given reservation is lost at run time due to type erasure. We cannot pattern match on a given reservation as on the following example:
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: > suffers from an inconvenient.

I think the adjective `inconvenient` should be replaced by the associated noun `inconvenience`.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 9 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 22 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: You are right. Changed.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 23]
```diff
@@ -0,0 +1,359 @@
...
+match may not be exhaustive.
+[warn] It would fail on the following inputs: (NormalSeat(_, _), Standing), (NormalSeat(_, _), VipTribune(_)), (Standing, NormalSeat(_, _)), (Standing, VipTribune(_)), (VipTribune(_), NormalSeat(_, _)), (VipTribune(_), Standing)
+[warn]     (userSeat, guestTicket) match {
+[warn]     ^
+```
+
+The compiler is concerned we might not be considering all possible combinations of seats within a reservation.   
+We are not considering for the cross cases (e.g. (Standing, NormalSeat)) because with the above tricks, we are certain that any created instance of Reservation will always respect the invariant. 
+The compiler isn’t. For it, a **single** Reservation[Seat] might contain tickets in different seating areas. We have to match on both tickets (user and guest) each of which, as far as the compiler goes, can be of any type (<code>Standing</code>, <code>VipTribune</code>, <code>NormalSeated</code>, <code>Seat</code>).  
+Because we (arguably) know more than the compiler in this case, we can tell him not to be concerned with the non-exhaustive pattern match with an annotation (chapter 15.5 at <a id="mmds-ref" href="#programming-in-scala">[1]</a>):
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md:300]: > Because we (arguably) know more than the compiler in this case

hahahahaha really like this statement!

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 24]
```diff
@@ -0,0 +1,359 @@
...
+  userTicket: Ticket[VipTribune],
+  guestTicket: Ticket[VipTribune]
+) extends Reservation[VipTribune]
+```
+
+On the other hand, all the type information is carried on at run-time. We can for example pattern match with confidence:  
+
+```scala
+def guideUsersToSeat(reservation: Reservation[Seat]): String = reservation match {
+  case ReservationStanding(_, _) => s"Go to gate 1."
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: No need for String interpolation here. I only raised this because you don't have that here:
> case _: Reservation[Standing.type] => "Go to gate 1. You and your guest must stand."

and 
> case (_: Standing.type, _: Standing.type ) => "Go to gate 1"

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 25]
```diff
@@ -0,0 +1,359 @@
...
+  case ReservationSeated(userTicket, guestTicket) => s"Go to gate 2. You and your guest are in a seats ${userTicket.seat.column} and ${guestTicket.seat.column} respectively "
+  case ReservationVipTribune(userTicket, _) => s"Go to gate 3. You are on tribune ${userTicket.seat.tribuneId}"
+  }
+```
+
+And because none of our sub-typed Reservation classes has tickets with seats of different types, the invariant is always assured.  
+
+This post is also on the author's blog at []()https://aerodatablog.wordpress.com/
+
+*Do you know any other approach ? I would be happy to hear it. Just write on the comment section.*
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please remove the extra whitespace before the question mark.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 8 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md:83]: I kinda agree.
For some reason I get an error when I do:
` case (Standing, Standing) => ......`
Did not try to understand why, just changed and moved on. (even though I could indeed change the other 3 cases to use _: type.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 21 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: Good eye. Thanks.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 24 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: Good eye.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 25 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 10 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: I always try to write in a way I think will be easier for people to understand (even though this is a easy topic)
For me, `can pretend to be` is more intuitive .

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 7 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md:66]: I tried in different ways and could not get one to violate that. 
If you find a way do tell me, so that we can see the problem and maybe find a solution.

File: _posts/2018-12-04-imposing-invariants-in-scala.md [hunk 26]
```diff
@@ -0,0 +1,359 @@
...
+sealed trait Reservation[+SeatType <: Seat]
+
+case class ReservationSeated(
+  userTicket: Ticket[NormalSeat],
+  guestTicket: Ticket[NormalSeat]
+) extends Reservation[NormalSeat]
+
+case class ReservationStanding(
+  userTicket: Ticket[Standing.type],
+  guestTicket: Ticket[Standing.type ]
```
Comment (sdsgomes) [inline _posts/2018-12-04-imposing-invariants-in-scala.md]: Please remove the extra whitespaces after `Standing.type`.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 14 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: DSL is jargon that we should avoid. This is not for experts which already have a precise idea of what a DSL is.
I could use a verb other than `discern`. `understand` is not suitable. Any suggestions ?

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 15 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: **First point**
`Regarding` has different semantics than `From`.
`From` gives the feeling of `as a consequence of` which `Regarding` does not. 

**Second point**
I can not point or reference the English grammar rule that states the current form to be the correct. Still, I feel it is the correct syntax.  
Perhaps you can ? :)

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 26 above)
Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: done

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 8 above)
Comment (Amuxix) [reply _posts/2018-12-04-imposing-invariants-in-scala.md:83]: You need to add the back ticks (`) otherwise it will try to unapply with the variable named Standing. If you copy the bit on the comment above it should work.

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 14 above)
Comment (sdsgomes) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: It's not the verb, it's the sentence construction. Perhaps this:
> The following diagram is useful for discerning the relation above and understanding the explanations that follow.

Does this seem good?

Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: Perfect. Will incorporate.

Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: Incorporated. Thanks

File: _posts/2018-12-04-imposing-invariants-in-scala.md (hunk 15 above)
Comment (sdsgomes) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: **First point**
I didn't understand you wanted a consequential effect. I'd still remove the "the" before "user ticket" and add a comma before "it thinks":
> From user ticket on line 2, it thinks it is VipTribune.

**Second point**
I didn't find any grammatical rule about that, but orally I think it sounds better the way I suggested. Still, this isn't an issue for me, just a suggestion.

Comment (cmhteixeira) [reply _posts/2018-12-04-imposing-invariants-in-scala.md]: **First point** Done

**Review by sdsgomes — APPROVED** [review]
